internal/transport/http2_client_test.go

```diff
@@ -0,0 +1,187 @@
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+ */
+package transport
+
+import (
+	"context"
+	"errors"
+	"net"
+	"testing"
+	"time"
+
+	"golang.org/x/net/http2"
+	"google.golang.org/grpc/resolver"
+)
+
+// ClientPreface is the HTTP/2 client preface string.
+var ClientPreface = []byte(http2.ClientPreface)
+
+var framerWindowUpdateString = "This is a framer window update test case string"
+
+var framerWriteSettingsString = "This is a framer write setting test case string"
+
+type clientPrefaceConn struct {
+	net.Conn
+}
+
+type clientPrefaceLengthConn struct {
+	net.Conn
+}
+
+type framerWriteSettingsConn struct {
+	net.Conn
+}
+
+type framerWindowUpdateConn struct {
+	net.Conn
+}
+
+func (cp *clientPrefaceConn) Write(b []byte) (n int, err error) {
+	if string(b) == string(ClientPreface) {
+		return 0, errors.New("preface write error")
+	}
+
+	// Normally write the bytes if they don't match the ClientPreface
+	return cp.Conn.Write(b)
+}
+
+func (cp *clientPrefaceConn) Close() error {
+	return cp.Conn.Close()
+}
+
+func dialerClientPrefaceWrite(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &clientPrefaceConn{Conn: conn}, nil
+}
+
+func (cpl *clientPrefaceLengthConn) Write(b []byte) (n int, err error) {
+	if string(b) == string(ClientPreface) {
+		incorrectPreface := "INCORRECT PREFACE\r\n\r\n"
+		n, err = cpl.Conn.Write([]byte(incorrectPreface))
+		return n, err
+	}
+	// Normally write the bytes if they don't match the ClientPreface
+	return cpl.Conn.Write(b)
+
+}
+
+func dialerClientPrefaceLength(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &clientPrefaceLengthConn{Conn: conn}, nil
+}
+
+func (fws *framerWriteSettingsConn) Write(b []byte) (n int, err error) {
+	if string(b) == framerWriteSettingsString {
+		return 0, errors.New("intended test case string detected: forced write error")
+	}
+	framerValue := 9
+	n, err = fws.Conn.Write(b)
+	// Compare the number of bytes written with the framer value
+	if n == framerValue {
+		return 0, errors.New("Framer write setting error")
+	}
+
+	return n, err
+}
+
+func dialerFramerWriteSettings(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &framerWriteSettingsConn{Conn: conn}, nil
+}
+
+func (fwu *framerWindowUpdateConn) Write(b []byte) (n int, err error) {
+	if string(b) == framerWindowUpdateString {
+		return 0, errors.New("intended test case string detected: forced write error")
```

Comment:
`intended test case string detected: forced write error` -> `force error for framerWindowUpdate`

Reply:
changes done